### PR TITLE
add CEXL docs

### PR DESCRIPTION
**Pull request by mandarjog**
Add md documentation for cexl.

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/264)

### CODE REVIEW

File: pkg/expr/expr.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Expression Language   (CEXL)
+
+Istio (mixer) configuration uses an expression language to evaluate selectors and input maps.
+The syntax of the expression language is C-like.
+
+An expression maps typed Attribute Values to a Value. 
+
+```
+f(attrValue0, attrValue1, ..., constant0, constat1, ...) --> Value
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: s/constat1/constant1/g

(this is copy/pasted elsewhere)

File: pkg/expr/expr.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+```
+f(attrValue0, attrValue1, ..., constant0, constat1, ...) --> Boolean
+```
+### Examples of selector expressions
+
+1. select all traffic that is destined for service1 and source user is admin.
+`target.service == "service1.ns1.cluster" && source.user == "admin"`
+2. select all requests that did not result in a error.
+`response.status_code < 400`
+3. select requests which were initially send to host "aaa".
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: s/send/sent/

File: pkg/expr/expr.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+Istio (mixer) configuration uses an expression language to evaluate selectors and input maps.
+The syntax of the expression language is C-like.
+
+An expression maps typed Attribute Values to a Value. 
+
+```
+f(attrValue0, attrValue1, ..., constant0, constat1, ...) --> Value
+```
+
+## Predicates
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: call this `Selector Expressions` to match the `Mapping Expressions` header and example below.

File: pkg/expr/expr.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+1. evaluates to request.size if available, otherwise 200
+`request.size| 200  --> int`
+2. evaluated to request.user if available, otherwise "nobody"
+`request.user | "nobody" --> string`
+
+Expression may use any valid attributes
+
+## Grammar
+Expression language uses a subset of C like expressions.
+The front end uses golang lexer and parser. The AST produced by the parser is converted to a simpler AST that only consists of Functions, Variables and Constants. Constants are implcitly typed. Variable names are treated as attribute name and therefore inherit type from the attribute definition. Function return values are either statically typed like `EQ(a, b ) --> Boolean` or dynamically typed like `OR(T a, T b) --> T`.
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: s/implcitly/implicitly/

File: pkg/expr/expr.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+Given a vocabulary every expression is validated and resolved to a type.
+Selectors __must__ resolve to a boolean value. Mapping expression should resolve to the type they are mapping into.   
+## Functions
+Expression language is extended by implementing the [Func interface](https://github.com/istio/mixer/blob/master/pkg/expr/func.go#L25). It is a small but growing list of functions.
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: can we link the `growing list` to the source file as well?

File: pkg/expr/expr.md (hunk 4 above)
Comment (douglas-reid) [inline pkg/expr/expr.md]: can we link `attribute definition` to the descriptor/vocabulary definition?

File: pkg/expr/expr.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+1. evaluates to request.size if available, otherwise 200
+`request.size| 200  --> int`
+2. evaluated to request.user if available, otherwise "nobody"
+`request.user | "nobody" --> string`
+
+Expression may use any valid attributes
+
+## Grammar
+Expression language uses a subset of C like expressions.
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: above, we used a hyphen for `C-like`. we should be consistent.

File: pkg/expr/expr.md (hunk 4 above)
Comment (douglas-reid) [inline pkg/expr/expr.md]: add link to golang lexer and golang parser.

Comment (douglas-reid) [inline pkg/expr/expr.md]: change to: "Variable names are treated as attribute name*s*... from the attribute definition*s*"

File: pkg/expr/expr.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
+# Expression Language   (CEXL)
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: Do we want/need to provide some hint as to where the C in CEXL comes from?

File: pkg/expr/expr.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+Mapping expressions map attribute values to a typed value.
+
+### Examples of mapping expressions
+
+1. evaluates to request.size if available, otherwise 200
+`request.size| 200  --> int`
+2. evaluated to request.user if available, otherwise "nobody"
+`request.user | "nobody" --> string`
+
+Expression may use any valid attributes
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: does this statement belong up top (universally-applicable)?

it should also end in a '.'

File: pkg/expr/expr.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+1. select all traffic that is destined for service1 and source user is admin.
+`target.service == "service1.ns1.cluster" && source.user == "admin"`
+2. select all requests that did not result in a error.
+`response.status_code < 400`
+3. select requests which were initially send to host "aaa".
+`request.header["X-FORWARDED-HOST"] == "aaa"`
+4. select all traffic that is destined for namespace ns1 and source user is admin.
+`target.service == "*.ns1.cluster" && source.user == "admin"`
+
+## Mapping expressions
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: Should `Mapping expressions` be the first section (as `Selector expressions` are just a special case of `Mapping expressions`) ?

File: pkg/expr/expr.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+`request.user | "nobody" --> string`
+
+Expression may use any valid attributes
+
+## Grammar
+Expression language uses a subset of C like expressions.
+The front end uses golang lexer and parser. The AST produced by the parser is converted to a simpler AST that only consists of Functions, Variables and Constants. Constants are implcitly typed. Variable names are treated as attribute name and therefore inherit type from the attribute definition. Function return values are either statically typed like `EQ(a, b ) --> Boolean` or dynamically typed like `OR(T a, T b) --> T`.
+
+Given a vocabulary every expression is validated and resolved to a type.
+Selectors __must__ resolve to a boolean value. Mapping expression should resolve to the type they are mapping into.
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: what happens if mapping expressions don't resolve to the declared type?  is `should` really `must` ?

File: pkg/expr/expr.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+## Grammar
+Expression language uses a subset of C like expressions.
+The front end uses golang lexer and parser. The AST produced by the parser is converted to a simpler AST that only consists of Functions, Variables and Constants. Constants are implcitly typed. Variable names are treated as attribute name and therefore inherit type from the attribute definition. Function return values are either statically typed like `EQ(a, b ) --> Boolean` or dynamically typed like `OR(T a, T b) --> T`.
+
+Given a vocabulary every expression is validated and resolved to a type.
+Selectors __must__ resolve to a boolean value. Mapping expression should resolve to the type they are mapping into.   
+## Functions
+Expression language is extended by implementing the [Func interface](https://github.com/istio/mixer/blob/master/pkg/expr/func.go#L25). It is a small but growing list of functions. 
+
+Typechecker ensures that all functions are called with the correct number of arguments of the correct type. The effective type of an expression is inferred by processing its AST.
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: What is `Typechecker`?  It seems irrelevant to the consumer of this document. This might be clearer just as something like:

```
Mixer's CEXL engine will ensure that all functions are called with the correct number and type of arguments.
```

File: pkg/expr/expr.md (hunk 5 above)
Comment (douglas-reid) [inline pkg/expr/expr.md]: Nit: Missing `The` or `Istio's` at beginning of sentence.

File: pkg/expr/expr.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+### Examples of mapping expressions
+
+1. evaluates to request.size if available, otherwise 200
+`request.size| 200  --> int`
+2. evaluated to request.user if available, otherwise "nobody"
+`request.user | "nobody" --> string`
+
+Expression may use any valid attributes
+
+## Grammar
```
Comment (douglas-reid) [inline pkg/expr/expr.md]: should we write out a full grammar in this section, to be explicit?  Or rename this section to something else?

File: pkg/expr/expr.md (hunk 4 above)
Comment (douglas-reid) [inline pkg/expr/expr.md]: probably should fully spell out Abstract Syntax Tree the first time we use it (and link to wiki or something describing it).

File: pkg/expr/expr.md (hunk 12 above)
Comment (douglas-reid) [inline pkg/expr/expr.md]: somewhere, it would be nice to have a fully-specified language definition for quick reference

File: pkg/expr/expr.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+4. select all traffic that is destined for namespace ns1 and source user is admin.
+`target.service == "*.ns1.cluster" && source.user == "admin"`
+
+## Mapping expressions
+Mapping expressions map attribute values to a typed value.
+
+### Examples of mapping expressions
+
+1. evaluates to request.size if available, otherwise 200
+`request.size| 200  --> int`
```
Comment (ZackButcher) [inline pkg/expr/expr.md]: add a space between `space` and the pipe.

File: pkg/expr/expr.md (hunk 4 above)
Comment (ZackButcher) [inline pkg/expr/expr.md]: Remove the extra space in `EQ(a, b ) --> Boolean` => `EQ(a, b) --> Boolean`

File: pkg/expr/expr.md [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+2. evaluated to request.user if available, otherwise "nobody"
+`request.user | "nobody" --> string`
+
+Expression may use any valid attributes
+
+## Grammar
+Expression language uses a subset of C like expressions.
+The front end uses golang lexer and parser. The AST produced by the parser is converted to a simpler AST that only consists of Functions, Variables and Constants. Constants are implcitly typed. Variable names are treated as attribute name and therefore inherit type from the attribute definition. Function return values are either statically typed like `EQ(a, b ) --> Boolean` or dynamically typed like `OR(T a, T b) --> T`.
+
+Given a vocabulary every expression is validated and resolved to a type.
```
Comment (ZackButcher) [inline pkg/expr/expr.md]: What are the type resolution rules? e.g. How do we resolve email vs string?

Comment (ZackButcher) [inline pkg/expr/expr.md]: Given a vocabulary<b>,</b> every ...

**Comment by rshriram** [comment]
Can this doc be either moved to istio/istio repo or atleast can we link from istio/istio to this doc?
We are trying to collate all docs there.

**Comment by geeknoid** [comment]
Can we close this PR?
